### PR TITLE
helm: allow annotations to be set for preflight resources

### PR DESCRIPTION
**Pull request by bradwhitfield**
Currently preflight resources, except the serviceaccount, do not allow annotations to be configured at the top level of the resources. This patch adds support for adding annotations.

Adding annotations at the top level of resources (not just in pod templates) will allow for use cases like ArgoCD and helm hooks.

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

```release-note
helm: allow annotations to be set for preflight resources
```

### CODE REVIEW

**Comment by gandro** [comment]
Thanks for the PR! Travis is failing because some files have not been regenerated:

```
HINT: to fix this, run 'make -C Documentation update-helm-values'
```

**Review by gandro — APPROVED** [review]
Thanks! PR and motivation look solid overall. There is an issue with CI (see previous comment) that needs to be fixed, but otherwise this looks good to me

**Comment by bradwhitfield** [comment]
Thanks, @gandro! I've been focused on getting the `diff` command to work in the `Preflight Clusterrole Check` that I missed that. It occurred to me that maybe a better approach for this PR is to include the annotations at the top levels of all resources. It does make the diff easier and I feel like a more consistent experience. Any reason not to just add annotations to everything?

**Review by youngnick — APPROVED** [review]
PR LGTM once the CI is sorted, nice work @bradwhitfield.

**Comment by gandro** [comment]
> It occurred to me that maybe a better approach for this PR is to include the annotations at the top levels of all resources. It does make the diff easier and I feel like a more consistent experience. Any reason not to just add annotations to everything?

I'm not sure I entirely understand your proposal. Would the idea be to have global annotations value that applies to all resources (preflight and regular deployments?). 

I think that's also a fine approach, though there is always a danger that contributors who add new resources might miss that they need to make the annotations customizable. Maybe we should have a CI or lint check for that if we decide to annotate all resources.

**Comment by gandro** [comment]
> So this is part of the diff of what I'm currently working on. I've created an annotations parameter for all of the various templates, if that makes sense.

Ah, thanks for the details! Yeah I think that does make sense too

**Comment by bradwhitfield** [comment]
Apologies for the delay. I've pushed what I came up with. While doing this, I had a couple of notes and questions I wanted to bring up. I can make any changes as needed based on feedback.

* `tls-cronjob/job.yaml` files already had annotations you can control. My thinking is that auto tls should have some sort of annotations across all the resources, so I’m curious to get feedback on how I handled those. For `tls-provided` resources, I just use the appropriate top-level annotations since you wouldn’t need them to be regenerated in the same way as helm or cert-manager might.
* There were a few files that I wasn’t sure the best approach for. Pretty much every file not under `template/` folder, I left alone. In my mind, they could either be annotated with the same annotations as the cilium-agent, or they could have unique annotation values. I was initially thinking the first option, but again was curious to get thoughts on that.
* In some places where it would have rendered and empty `annotations` before, I made conditionally generate that field. I can undo this if needed.
* Do you want it to be consistent wether labels or annotations are first?

File: Documentation/helm-values.rst
Comment (gandro) [inline Documentation/helm-values.rst:2527]: nit: Please squash this into the previous commit

File: install/kubernetes/cilium/README.md
Comment (gandro) [inline install/kubernetes/cilium/README.md:63]: Same as above: Please squash into the previous commit. That way we can use "git bisect" to build without errors

File: install/kubernetes/cilium/templates/cilium-envoy/service.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-envoy/service.yaml]: This looks correct to me

File: install/kubernetes/cilium/templates/cilium-nodeinit/serviceaccount.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-nodeinit/serviceaccount.yaml]: This should be `.Values.nodeinit.annotations` in the second expression

File: install/kubernetes/cilium/templates/clustermesh-apiserver/tls-certmanager/admin-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/tls-certmanager/admin-secret.yaml]: I think we probably can just use the overall clustermesh-api-server annotations here, i.e. `.Values.clustermesh.annotations`, same for all other clustermesh TLS resources

File: install/kubernetes/cilium/templates/hubble/tls-certmanager/relay-client-secret.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/hubble/tls-certmanager/relay-client-secret.yaml]: I think we could just use `.Values.hubble.annotations` for all TLS related objects (i.e. templates/hubble/tls-*). It would make it a bit easier I think

**Comment by gandro** [comment]
This is excellent work! I'll leave a few minor nits in inline comments.

>     * `tls-cronjob/job.yaml` files already had annotations you can control. My thinking is that auto tls should have some sort of annotations across all the resources, so I’m curious to get feedback on how I handled those. For `tls-provided` resources, I just use the appropriate top-level annotations since you wouldn’t need them to be regenerated in the same way as helm or cert-manager might.

The current solution looks good to me, though I think it might be simpler to use the top-level annotations for all TLS resources, regardless if provided or not.

>     * There were a few files that I wasn’t sure the best approach for. Pretty much every file not under `template/` folder, I left alone. In my mind, they could either be annotated with the same annotations as the cilium-agent, or they could have unique annotation values. I was initially thinking the first option, but again was curious to get thoughts on that.

Yeah, I think we can always add more annotations if needed. From what I can tell, your current choice looks good to me.

>     * In some places where it would have rendered and empty `annotations` before, I made conditionally generate that field. I can undo this if needed.

Yeah, the if guards look good to me overall. I think that's a good solution.

>     * Do you want it to be consistent wether labels or annotations are first?

The PR is already pretty large. I think that is better done in a follow-up PR, but I don't think it's a must-have.

**Review by gandro — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-envoy/service.yaml
Comment (bradwhitfield) [reply install/kubernetes/cilium/templates/cilium-envoy/service.yaml]: Didn't meant to leave that in. I'll clean that up. Good catch.

File: install/kubernetes/cilium/README.md
Comment (bradwhitfield) [reply install/kubernetes/cilium/README.md:63]: So to make sure I'm clear, should I end up with two commits like these, or just one commit like the bottom one?
* helm: allow annotations to be set for preflight resources 
* helm: allow annotations to be set on nearly all resources

edit:

For the PR feedback, should that be in another commit, or would you like it amended to the `helm: allow annotations to be set on nearly all resources` commit?

**Comment by bradwhitfield** [comment]
Thank you for all the feedback! I'm just getting back from being out a few and will work on address this tomorrow.

**Comment by aditighag** [comment]
@bradwhitfield Feel free to move it out of draft once you've addressed all the review comments, and it's ready for review.

**Comment by bradwhitfield** [comment]
It's basically ready. I just need to rebase and clean up the git history but wanted to make sure I understood what it should  ideally look like based on [this comment](https://github.com/cilium/cilium/pull/27860#discussion_r1328466899).

File: install/kubernetes/cilium/README.md
Comment (gandro) [reply install/kubernetes/cilium/README.md:63]: I think just having the two commits that you outlined will yield a clean git history, that sounds good to me. No need to put the PR feedback in a separate commit, since GitHub can show us those diffs too (as long as you don't change the commits and rebase at the same time).

**Review by gandro — APPROVED** [review]
Code-wise this looks good to me now, thanks!

**Comment by bradwhitfield** [comment]
I think this is ready. I see that the [check failed](https://github.com/cilium/cilium/actions/runs/6318239572/job/17156809523?pr=27860), but I did amend the commit to fix that. It doesn't appear to have retriggered the build if I'm parsing things correctly. Is there a way I can trigger that?

**Review by aditighag — APPROVED** [review]
Approving for my codeowners

**Comment by aditighag** [comment]
Review was requested from sig-servicemesh, but no one was specifically assigned for reviews. /cc @cilium/sig-servicemesh 
I've marked the PR as ready for merge, but if you have any comments, feel free to post them on the PR.

**Comment by gandro** [comment]
Let's run CI as well, before we merge

**Comment by gandro** [comment]
All required tests have passed. The failing tests are not blockers:

- checkpatch is complaining about an overly long git commit subject, but that's not a hard requirement
- ci-aks timed out https://github.com/cilium/cilium/actions/runs/6323246632/job/17170451500
- ci-ipsec-uprade https://github.com/cilium/cilium/actions/runs/6323247394/job/17170446827 hit https://github.com/cilium/cilium/issues/27827

Marking again as ready to merge. There is an outstanding request from sig-servicemesh, but Nick (who is part of that team) has already reviewed.

**Comment by lmb** [comment]
sig-servicemesh hasn't approved yet?

File: install/kubernetes/Makefile
Comment (joestringer) [inline install/kubernetes/Makefile:47]: Why was this change made? It doesn't seem to be related to the goal of the PR.

Comment (joestringer) [reply install/kubernetes/Makefile:47]: @gandro @youngnick @aditighag I know this is a nit, but how is this falling through review?

Comment (aditighag) [reply install/kubernetes/Makefile:47]: It was probably an oversight. Based on my update, I only looked at the files for my CODEOWNERS (contributing), so I most likely didn't look at the changes in this file. 

`/install/kubernetes/ @cilium/sig-k8s @cilium/helm`

Comment (bradwhitfield) [reply install/kubernetes/Makefile:47]: That was not intentional. I had to fight a bit to get the makefiles to work on my computer. I eventually did. I don't remember this change specifically, but I probably commented it out for testing and didn't catch that I committed it. I can open a PR to fix that if you'd like.

File: install/kubernetes/cilium/templates/cilium-agent/serviceaccount.yaml
Comment (coro) [inline install/kubernetes/cilium/templates/cilium-agent/serviceaccount.yaml:19]: @bradwhitfield This is causing a duplicate `annotations` key if `.Values.serviceAccounts.cilium.annotations` is specified in the Helm release. Would you mind sending in a PR to fix this?

File: install/kubernetes/Makefile
Comment (gandro) [reply install/kubernetes/Makefile:47]: @bradwhitfield Yeah, a PR to fix it would be appreciated.
Also https://github.com/cilium/cilium/pull/27860#discussion_r1378867381 sounds like something we should address in a follow-up PR.

@joestringer I'm not sure how this slipped during review. It seems the [commit which did the change in question here](https://github.com/cilium/cilium/pull/27860/commits/dea1d189e46945f5f0cfffeae366f3e103ba9999) was pushed after I already acked the PR. I should have performed a re-review before running CI, I assumed the changes were mostly due to rebases. Sorry about that.

Comment (bradwhitfield) [reply install/kubernetes/Makefile:47]: I submitted #28929. The checks are passing and I believe I did the "fixes" correctly.

Comment (joestringer) [reply install/kubernetes/Makefile:47]: Thanks all :+1: My goal with starting the thread was to understand whether there's some underlying issue on some platforms that might have motivated this change in the first place (in which case maybe we can tweak the makefiles/scripts to be more platform-friendly?) and to get a better understanding of why unintentional changes may be introduced into the tree.

While in general I think we have a policy not to add commented code to the tree, it is ultimately up to the reviewers to enforce these constraints and I'd typically expect reviewers to collectively lay eyes on every line of code and have an understanding about how that code fits into the broader picture. In this case it seems like the submission slipped through both Brad's filter as he submitted the PR and the reviewer's filters as they accepted the PR. 

There's always a small chance of this sort of thing happening, and I understand that contributors can make mistakes or have a slip of oversight, particularly when changes are introduced later in the review cycle. Given the brief retrospective we've collectively done here, I don't think there is a systematic issue to be concerned about beyond this PR; it's just a classic case of not realizing that there was another change that came in after initial review which we would usually not accept. Fortunately the consequences in this case are low.

I hope that this discussion has helped with sharing understandings and expectations around the PR submission process going forward. I appreciate the quick feedback & solution @bradwhitfield :pray:

**Pull request merged**
